### PR TITLE
🦊🐺🐭 Add connectors

### PR DESCRIPTION
**Pull request by dmaretskyi**
closes https://github.com/TrueFiEng/useDApp/issues/788
closes https://github.com/TrueFiEng/useDApp/issues/520
closes https://github.com/TrueFiEng/useDApp/issues/419
closes https://github.com/TrueFiEng/useDApp/issues/282

### TODO

- [x] Portis connector
- [x] Coinbase connector
- [x] MetaMask connector in config by default - backwards compat.
    - [x] By default - all connectors that do not need any user config parameters
- [x] When MetaMask changes network, the UI should update sooner
- [x] eip1193 - not working yet. Is it the same issue as above?
- [x] Accept readOnlyURLs in connectors for read-only connectors (ApiNode)
- [x] Use Event from dxos `onUpdate`
- [x] Check if `useConnectorNetwork` is needed
- [x] Reimplement `pollingInterval` logic
- [x] Consider deleting connectEagerly
- [x] Watch the bundle size - set the wallet packages as (optional) peer deps?

### CODE REVIEW

File: packages/core/src/providers/network/connector/connectorController.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { Connector } from "./connector";
+
+export class ConnectorController {
```
Comment (dmaretskyi) [inline packages/core/src/providers/network/connector/connectorController.ts:3]: should handle the lifecycle of the connector including chainID, accounts, block number etc.

Should have safeguards to protect from bugs in the connector implementation

File: packages/core/src/helpers/getSignerFromOptions.ts [hunk 2]
```diff
@@ -22,5 +22,7 @@ export const getSignerFromOptions = (
 
   const optionsSigner = options && 'signer' in options && options.signer
 
-  return privateKeySigner || mnemonicPhraseSigner || encryptedJsonSigner || optionsSigner || library?.getSigner()
+  const overriddenSigner = privateKeySigner || mnemonicPhraseSigner || encryptedJsonSigner || optionsSigner
```
Comment (mj426382) [inline packages/core/src/helpers/getSignerFromOptions.ts:25]: Revert changes

File: packages/core/src/providers/LocalMulticallProvider.tsx [hunk 3]
```diff
@@ -36,6 +36,14 @@ export function LocalMulticallProvider({ children }: LocalMulticallProps) {
     } else if (multicallAddresses && multicallAddresses[chainId]) {
       setLocalMulticallState(LocalMulticallState.Deployed)
     } else if (localMulticallState !== LocalMulticallState.Deploying) {
+      if (!(library instanceof providers.JsonRpcProvider)) {
+        throw new Error('You cannot send transaction without wallet')
```
Comment (mj426382) [inline packages/core/src/providers/LocalMulticallProvider.tsx:40]: Investigate removing

File: packages/core/src/providers/network/connector/impls/index.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export { WalletConnectConnector } from './walletConnect'
+export { MetamaskConnector } from './metamask'
+export { PortisConnector } from './portis'
```
Comment (mj426382) [inline packages/core/src/providers/network/connector/impls/index.ts:3]: Export defaultConnector

File: packages/example/package.json [hunk 5]
```diff
@@ -25,6 +25,8 @@
     "src"
   ],
   "dependencies": {
+    "@portis/web3": "^4.0.7",
```
Comment (mj426382) [inline packages/example/package.json:28]: Remove

File: packages/example/playwright/with-metamask.test.ts [hunk 6]
```diff
@@ -66,4 +66,110 @@ describe(`Browser: ${browserType.name()} with Metamask`, () => {
...
+
+    it('Check if all chains were loaded', async () => {
+      await page.goto(`${baseUrl}multichain`)
+
+      await waitForExpect(async () => {
+        expect(await page.isVisible(XPath.text('span', 'Chain id:', 4))).to.be.true
+        expect(await page.isVisible(XPath.text('span', 'Current block timestamp:', 4))).to.be.true
+        expect(await page.isVisible(XPath.text('span', 'Current difficulty:', 4))).to.be.true
+        expect(await page.isVisible(XPath.text('span', 'Current block:', 4))).to.be.true
+        expect(await page.isVisible(XPath.text('span', 'Ether balance:', 4))).to.be.true
```
Comment (mj426382) [inline packages/example/playwright/with-metamask.test.ts:91]: Remove

File: .github/workflows/CI.yml [hunk 7]
```diff
@@ -36,31 +36,34 @@ jobs:
...
           set -e
-          pnpm start &
+          LOCALHOST_URL='http://localhost:8545' pnpm start &
           xvfb-run --auto-servernum --server-args="-screen 0 1280x960x24" -- pnpm test:playwright
         env:
           MAINNET_URL: ${{ secrets.MAINNET_PROVIDER_URL }}
           DEBUG: usedapp*
         working-directory: packages/example
       - name: Run playwright example-next
+        # echo "NEXT_PUBLIC_LOCALHOST_URL=http://localhost:8545" > .env.local - couln't figure out how to set env variable the same as in previous section
```
Comment (rzadp) [inline .github/workflows/CI.yml:51]: Can be removed?

File: packages/connectors/coinbase-wallet/package.json [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  "main": "dist/src/index.js",
+  "types": "dist/src/index.d.ts",
+  "repository": "git@github.com:EthWorks/useDApp.git",
+  "author": "Ethworks",
+  "license": "MIT",
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@usedapp/core": "workspace:*",
```
Comment (rzadp) [inline packages/connectors/coinbase-wallet/package.json:13]: should be a peer?

File: packages/connectors/portis/package.json [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  "main": "dist/src/index.js",
+  "types": "dist/src/index.d.ts",
+  "repository": "git@github.com:EthWorks/useDApp.git",
+  "author": "Ethworks",
+  "license": "MIT",
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@usedapp/core": "workspace:*",
```
Comment (rzadp) [inline packages/connectors/portis/package.json:13]: should be peer?

File: packages/connectors/wallet-connect/package.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  "main": "dist/src/index.js",
+  "types": "dist/src/index.d.ts",
+  "repository": "git@github.com:EthWorks/useDApp.git",
+  "author": "Ethworks",
+  "license": "MIT",
+  "publishConfig": {
+    "access": "public"
+  },
+  "dependencies": {
+    "@usedapp/core": "workspace:*",
```
Comment (rzadp) [inline packages/connectors/wallet-connect/package.json:13]: should be peer?

File: packages/core/src/hooks/useEtherBalance.test.tsx [hunk 11]
```diff
@@ -75,7 +75,7 @@ describe('useEtherBalance', () => {
     )
     await waitForCurrent((val) => val !== undefined)
     expect(result.error).to.be.undefined
-    expect(result.current).to.eq(200)
+    expect(result.current).to.eq(100)
```
Comment (rzadp) [inline packages/core/src/hooks/useEtherBalance.test.tsx:78]: hmm should this be different?

File: packages/core/src/hooks/useEthers.test.tsx [hunk 12]
```diff
@@ -34,7 +34,7 @@ describe('useEthers', () => {
     expect(result.current.account).to.be.undefined
     expect(result.current.error).to.be.undefined
     expect(result.current.library).to.eq(network1.provider)
-    expect(result.current.active).to.be.true
+    expect(result.current.active).to.be.false
```
Comment (rzadp) [inline packages/core/src/hooks/useEthers.test.tsx:37]: hmm should this be false? Should we wait for it to be true?

File: packages/core/src/providers/network/connector/impls/walletConnect.ts [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  async connectEagerly(): Promise<void> {
+    await this.init()
+    try {
+      const chainId: string = await this.provider!.send('eth_chainId', [])
+      const accounts: string[] = await this.provider!.send('eth_accounts', [])
+      this.onUpdate?.({ chainId: parseInt(chainId), accounts })
+    } catch (e) {
+      console.log(e)
```
Comment (rzadp) [inline packages/core/src/providers/network/connector/impls/walletConnect.ts:37]: Consider console.debug? This is connectEagerly, so we kinda expect it to fail

File: packages/core/src/providers/network/connector/impls/testingWallet.ts [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+  async connectEagerly(): Promise<void> {
+    if (!this.provider) {
+      return
+    }
+    try {
+      const chainId: string = await this.provider!.send('eth_chainId', [])
+      this.onUpdate?.({ chainId: parseInt(chainId), accounts: [] })
+    } catch (error) {
+      console.log(error)
```
Comment (rzadp) [inline packages/core/src/providers/network/connector/impls/testingWallet.ts:29]: Consider console.debug? This is connectEagerly, so we kinda expect it to fail

File: packages/core/src/providers/network/connector/impls/portis.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    this.provider = new providers.Web3Provider(portis.provider)
+  }
+
+  async connectEagerly(): Promise<void> {
+    try {
+      await this.init()
+      const accounts: string[] = await this.provider!.send('eth_accounts', [])
+      this.onUpdate?.({ chainId: this.chainId, accounts })
+    } catch (e) {
+      console.log(e)
```
Comment (rzadp) [inline packages/core/src/providers/network/connector/impls/portis.ts:38]: Consider console.debug? This is connectEagerly, so we kinda expect it to fail

File: packages/core/src/providers/network/connector/impls/metamask.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  async connectEagerly(): Promise<void> {
+    await this.init()
+
+    if (!this.provider) {
+      return
+    }
+
+    const chainId: string = await this.provider!.send('eth_chainId', [])
```
Comment (rzadp) [inline packages/core/src/providers/network/connector/impls/metamask.ts:34]: No try-catch? Does it work fine in a browser without metamask?

File: packages/example-next/components/NavBar.tsx [hunk 17]
```diff
@@ -47,6 +47,10 @@ export function NavBar() {
               </>
             </StyledLink>
             <StyledLink href="/wallet-connect"> WalletConnect example </StyledLink>
+            <StyledLink href="/connector">
```
Comment (rzadp) [inline packages/example-next/components/NavBar.tsx:50]: connectorS ?

File: packages/example-next/package.json [hunk 18]
```diff
@@ -25,6 +29,7 @@
     "@types/styled-components": "^5.1.7",
     "eslint": "7.22.0",
     "eslint-config-next": "12.1.6",
+    "ethereum-waffle": "4.0.0-alpha.27",
```
Comment (rzadp) [inline packages/example-next/package.json:32]: normal release can be used now

File: packages/example-next/pages/connector/index.tsx [hunk 19]
```diff
@@ -0,0 +1,5 @@
+import { ConnectorPage } from '@usedapp/example'
```
Comment (rzadp) [inline packages/example-next/pages/connector/index.tsx:1]: ConnectorsPage?

File: packages/example/playwright/without-metamask.ts [hunk 20]
```diff
@@ -20,17 +26,103 @@ export const withoutMetamaskTest = (url: string) => {
...
+        await waitForExpect(async () => {
+          expect(await page.isVisible(XPath.text('span', 'Chain id:'))).to.be.true
+          expect(await page.isVisible(XPath.text('span', 'Current block:'))).to.be.true
+          expect(await page.isVisible(XPath.text('span', 'Current difficulty:'))).to.be.true
+          expect(await page.isVisible(XPath.text('span', 'Current block timestamp:'))).to.be.true
+        })
+      })
+    })
+
+    describe('Tokens', () => {
```
Comment (rzadp) [inline packages/example/playwright/without-metamask.ts:84]: We can remove it as an uncuccesful experiment

File: packages/example/src/components/SendEthForm/SendEthForm.tsx [hunk 21]
```diff
@@ -71,8 +71,8 @@ const InputComponent = () => {
 }
 
 export const SendEthForm = () => {
-  const { account } = useEthers()
-  const balance = useEtherBalance(account)
+  const { account, chainId } = useEthers()
+  const balance = useEtherBalance(account, { chainId })
```
Comment (rzadp) [inline packages/example/src/components/SendEthForm/SendEthForm.tsx:75]: Is this needed?

**Pull request closed without merging**